catalyst_voices_packages/catalyst_cardano_serialization/lib/src/rbac/role_based_access_control.dart

```diff
@@ -0,0 +1,354 @@
+import 'package:catalyst_cardano_serialization/catalyst_cardano_serialization.dart';
+import 'package:cbor/cbor.dart';
+import 'package:equatable/equatable.dart';
+
+/// Defines the X509 Role Based Access Control transaction metadata.
+final class RbacMetadata extends Equatable {
+  /// Un-ordered List of DER encoded x509 certificates.
+  final List<X509DerCertificate>? derCerts;
+
+  /// Un-ordered List of CBOR encoded C509 certificates
+  /// (or metadatum references).
+  //
+  // TODO(dtscalac): support C509CertInMetadatumReference
+  final List<C509Certificate>? cborCerts;
+
+  /// Ordered list of simple public keys that are registered.
+  final List<Ed25519PublicKey>? publicKeys;
+
+  /// Revocation list of certs being revoked by an issuer.
+  final List<CertificateHash>? revocationSet;
+
+  /// Set of role registration data.
+  final Set<RoleData>? roleDataSet;
+
+  /// The default constructor for [RbacMetadata].
+  const RbacMetadata({
+    this.derCerts,
+    this.cborCerts,
+    this.publicKeys,
+    this.revocationSet,
+    this.roleDataSet,
+  });
+
+  /// Deserializes the type from cbor.
+  factory RbacMetadata.fromCbor(CborValue value) {
+    final map = value as CborMap;
+    final derCerts = map[const CborSmallInt(10)] as CborList?;
+    final cborCerts = map[const CborSmallInt(20)] as CborList?;
+    final publicKeys = map[const CborSmallInt(30)] as CborList?;
+    final revocationSet = map[const CborSmallInt(40)] as CborList?;
+    final roleDataSet = map[const CborSmallInt(50)] as CborList?;
```

Comment:
Yes, it will fail if somebody will incorrectly set key: 50 as different cbor value, i.e. CborString. Then it should fail and you should fix your data.

Reply:
Should we then consider creating a fallible initializer and catching & throwing an exception in case data is not correct?

Reply:
It's extra effort and serialization code then becomes more complex but the user of the package might get some extra info.

The most popular serialization frameworks, i.e. json_serializable don't do it